### ISSUE TITLE
Enhancement: [only-throw-error] allow re-throwing a caught error

### ISSUE THREAD

**Issue by ronami** [typescript-eslint/typescript-eslint#10376]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/only-throw-error

### Description

`only-throw-error` (and `prefer-promise-reject-errors`) disallow throwing non-error values. `only-throw-error` (and with #10375 `prefer-promise-reject-errors` too) can be configured to allow throwing `unknown` and `any` types in addition to error values.

It's necessary to enable `allowThrowingUnknown` so re-throwing a cached error is valid (or `any` in some `tsconfig.json` configurations):

```ts
function test() {
  try {
    // ...
  } catch (error) {
    if (x) {
      throw new Error()
    }

    // this would otherwise fail
    throw error;
  }
}
```

While configuring `allowThrowingUnknown` makes the above pass, it allows potentially unsafe throwing of an `unknown` type:

```ts
function foo(): unknown {
  // not necessarily returns an `Error` object
}

function test() {
  // this should fail
  throw foo();
}
```

This issue proposes an additional option, a stricter form of `allowThrowingUnknown,` which will allow re-throwing a caught error.

_This is relevant both to [`only-throw-error`](https://typescript-eslint.io/rules/only-throw-error) and [`prefer-promise-reject-errors`](https://typescript-eslint.io/rules/prefer-promise-reject-errors)._

### Fail

```TypeScript
function foo(): unknown {
  // not necessarily returns an `Error` object
}

function test() {
  // this should fail
  throw foo();
}
```

### Pass

```TypeScript
function test() {
  try {
    // ...
  } catch (error) {
    if (x) {
      throw new Error()
    }

    throw error;
  }
}
```

**Comment by bradzacher** [comment]
Counter argument to your request - it's no safer really and can lead to ugly behaviour:

```ts
try {
  // imagine this is in a 3rd party library
  throw 'lol oops';
} catch (e) {
  throw e;
}
```

**Comment by ronami** [comment]
> Counter argument to your request - it's no safer really and can lead to ugly behaviour:
> 
> ```ts
> try {
>   // imagine this is in a 3rd party library
>   throw 'lol oops';
> } catch (e) {
>   // this seems safe and wouldn't error in the rule despite it being unsafe.
>   throw e;
> }
> ```

Yes, it's not as safe as using the rule without any options, but I think it is safer than having `allowThrowingUnknown` (or `allowThrowingAny` in some cases) enabled. It allows a specific error handling pattern rather than allowing throwing any `unknown` type, regardless of where it came from.

I don't know if it's too niche, though.

**Comment by kirkwaiblinger** [comment]
I am +1. 

Partially handling/logging, then rethrowing, an error is a valid pattern that does happen. In other languages (for more technical reasons to do with stack traces) there is even a dedicated `throw;` or `raise` statement (taking no operand) for this purpose, and if JS had such a thing, we probably wouldn't think twice about the validity of using it. Propagating a potentially non-error exception is a different - and much less dev-friendly to avoid - thing than creating one. 

This is largely analogous to https://github.com/typescript-eslint/typescript-eslint/issues/8538, where propagating a bad `void` is treated differently than creating a bad `void`.

This probably makes sense as an option here too.

**Comment by JoshuaKGoldberg** [comment]
I like this idea in theory, but don't think it can be done well practically. Rules targeting specific syntax uses are brittle to refactors. There's no real difference in developer intent between re-throwing _directly_ inside a `catch` vs. re-throwing _indirectly_, such as inside a helper:

```ts
function doDangerousWork(work: () => void) {
  try {
    work()
  } catch (error) {
    logAndRethrow(error);
  }
}

function logAndRethrow(error: unknown) {
  console.log("oh no", error);
  throw error;
}
```

For that reason I'm -1 on this issue. My suggestion would be to reuse the existing `allowThrowingAny` and `allowThrowingUnknown` options as described in the OP.

**Comment by kirkwaiblinger** [comment]
> I like this idea in theory, but don't think it can be done well practically. Rules targeting specific syntax uses are brittle to refactors. There's no real difference in developer intent between re-throwing _directly_ inside a `catch` vs. re-throwing _indirectly_, such as inside a helper:
> 
> ```ts
> function doDangerousWork(work: () => void) {
>   try {
>     work()
>   } catch (error) {
>     logAndRethrow(error);
>   }
> }
> 
> function logAndRethrow(error: unknown) {
>   console.log("oh no", error);
>   throw error;
> }
> ```
> 
> For that reason I'm -1 on this issue. My suggestion would be to reuse the existing `allowThrowingAny` and `allowThrowingUnknown` options as described in the OP.

I get what you're saying here, but am not swayed. The trouble is `allowThrowingUnknown: false` is basically just unusable for a large class of valid patterns as-is.

Why should a user who is trying to do this
```ts
function doDangerousWork(work) {
  try {
    work()
  } catch (error: unknown) {
    console.error('critical thing failed!')
    throw error;
  }
}
```
be forced to pick between working around the option for an easily-detectable valid pattern
```ts
function doDangerousWork() {
  try {
    work()
  } catch (error: unknown) {
    console.error('critical thing failed!')

    // yuck
    if (error instanceof Error) {
      throw error;
    }
    throw new Error("error wasn't an error", { cause: error });
  }
}
```
or disabling the option and allowing unrelated, bad code
```ts
function doDangerousWork(args: unknown) {
    throw args;
}
```
?

In effect, it _encourages_ eslint-disable-averse users to write a helper to rethrow, which is objectively _worse_ code.
```ts
function rethrow(error: unknown): never {
  // eslint-disable-next-line @typescript-eslint/only-throw-error -- utility to allow rethrowing without linter complaints
  throw error;
}

function doDangerousWork(work) {
  try {
    work()
  } catch (error: unknown) {
    console.error('critical thing failed!')
    rethrow(error);
  }
}
```


***

Note that another pattern that this messes up is the ability to do conditional behavior in a `finally` block based on whether an exception was encountered.

```ts
function doDangerousWork() {
   let didThrow = false;
   try { 
     work();
   } catch (x) {
      didThrow = true;
      // why do i have to do this when i am making no attempt to 
      // handle the error or throw a new exception? I'm just trying
      // to observe whether there was an exception.
      if (error instanceof Error) {
         throw error;
      } else {
         throw new Error('error wasnt an error', { cause: error });
      }
   } finally {
     doConditionalCleanup(didThrow);
   }
}
```

***

I think there's a meaningful distinction between whether an exception is being _created_ or whether it's being rethrown because the language forces you to, and I don't think we should put the responsibility to _repair_ a possibly-invalid exception during an obviously-detectable rethrowing pattern. Especially since exceptions _never_ have valid types in a catch block.

Taking the counterargument ad absurdum, we could even say this line should be flagged since it possibly throws an exception that isn't an error
```ts
function doWork() {
   doesntThrow();
}
```
and therefore every function call in the program should be fixed to 
```ts
function doDangerousWork() {
  try { 
    doesntThrow();
  } catch (e) { 
    if (e instanceof Error) {
      throw e;
    } else {
      throw new Error("non-error Exception", { cause: e });
    }
  }
   
}
```

**Comment by kirkwaiblinger** [comment]
So, in conclusion, I'd rather take the downside of having a very, very natural and understandable syntax heuristic (which is probably the _only_ correct syntax anyway to get technically clean stack traces), than the downside of flagging on every rethrow.

**Comment by ronami** [comment]
Would it make sense to have this option be on by default?

I've enabled both [`only-throw-error`](https://typescript-eslint.io/rules/only-throw-error) and [`prefer-promise-reject-errors`](https://typescript-eslint.io/rules/prefer-promise-reject-errors) on a couple of projects, and I had to add an `eslint-disable-next-line` comment on top of several of such cases in each project.

This is similar to the original plan to have the option added in https://github.com/typescript-eslint/typescript-eslint/issues/8538 enabled by default in a future major. I think this is a significant pain point in using the rule on a somewhat large codebase.

**Comment by JoshuaKGoldberg** [comment]
🤔 maybe enabled in recommended, disabled in strict? 🤷 I wouldn't be upset by just always having it enabled by default.

**Comment by kirkwaiblinger** [comment]
+1 to just enabled by default always.

@ronami are you asking because you plan to work on this issue? I started working on this today, so I figure I'll share my notes if so; my branch is here: https://github.com/typescript-eslint/typescript-eslint/compare/main...kirkwaiblinger:typescript-eslint:only-throw-error-rethrow

**Comment by ronami** [comment]
> +1 to just enabled by default always.
> 
> @ronami are you asking because you plan to work on this issue? I started working on this today, so I figure I'll share my notes if so; my branch is here: [main...kirkwaiblinger:typescript-eslint:only-throw-error-rethrow](https://github.com/typescript-eslint/typescript-eslint/compare/main...kirkwaiblinger:typescript-eslint:only-throw-error-rethrow)

Oh, not at all. Please go ahead and continue. This will be helpful to me too on my own projects. Thanks for taking this!

I really enjoyed reading your reasoning in this discussion. I think you've explained why this option is important really well (especially on the difference between creating a bad throw and propagating an existing one). I didn't add any comments because I didn't have anything useful to add, thanks! ❤️
